src/gpu/nvidia/sycl_cuda_memory_storage_helper.hpp

```diff
@@ -0,0 +1,96 @@
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
+* Copyright 2022 Codeplay Software Limited
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef GPU_NVIDIA_SYCL_CUDA_MEMORY_STORAGE_HELPER_HPP
+#define GPU_NVIDIA_SYCL_CUDA_MEMORY_STORAGE_HELPER_HPP
+
+#include <optional>
+#include "gpu/nvidia/sycl_cuda_scoped_context.hpp"
+#include "sycl/sycl_memory_storage.hpp"
+
+namespace dnnl {
+namespace impl {
+namespace gpu {
+namespace nvidia {
+
+#define CTX_IN_SYCL_MEMORY(arg) \
+    sycl_memory_arg_t<::sycl::access::mode::read>( \
+            static_cast<sycl::sycl_memory_storage_base_t *>( \
+                    &CTX_IN_STORAGE(arg)), \
```

Comment:
`CTX_IN_STORAGE` can return `memory_storage_t::empty_storage()` that cannot be casted to `sycl_memory_storage_base_t`.

I suggest to pass whatever `CTX_(IN|OUT)_STORAGE` returns to the constructor. If it's an empty storage then
`sycl_memory_arg_t` will have an empty state. In order to check whether it's empty `empty()` member function will be introduced. With this, we can simplify places where arguments are optional, e.g. workspace for training.

Example:
```cpp
sycl_memory_arg_t<::sycl::access::mode::read>(&CTX_IN_STORAGE(arg), cgh)
```
```cpp
sycl_memory_arg_t(memory_storage_t *mem, ::sycl::handler &cgh) {
    // sycl_memory_arg_t will be empty.
    if (mem->is_null()) return;
    auto *sycl_storage = static_cast<sycl::sycl_memory_storage_base_t *>(mem);
    // ...
}
```
```cpp
bool empty() const { return !raw_ptr_ && !acc_.has_value(); }
```

Reply:
done

Reply:
Thanks!